src/encoding.nr

```diff
@@ -311,68 +205,99 @@ pub(crate) fn encode_message_extension(w: Field) -> EncodedWitness {
     base4_encoded_slices[10] = BASE4_ENCODE_9BIT_TABLE[s[10]];
     base4_encoded_slices[11] = BASE4_ENCODE_3BIT_TABLE[s[11]];
 
-    // 42
-    let mut bits19to61 = base4_encoded_slices[10];
-    bits19to61 *= BASE4_POWERS[9];
-    bits19to61 += base4_encoded_slices[9];
-    bits19to61 *= BASE4_POWERS[8];
-    bits19to61 += base4_encoded_slices[8];
-    bits19to61 *= BASE4_POWERS[8];
-    bits19to61 += base4_encoded_slices[7];
-    bits19to61 *= BASE4_POWERS[8];
-    bits19to61 += base4_encoded_slices[6];
-
-    let mut bits8to61 = bits19to61 * BASE4_POWERS[3];
-    bits8to61 += base4_encoded_slices[5];
-    bits8to61 *= BASE4_POWERS[8];
-    bits8to61 += base4_encoded_slices[4];
-
-    let low1 = base4_encoded_slices[0];
-
-    let bits1to8 = base4_encoded_slices[1]
-        + base4_encoded_slices[2] * BASE4_POWERS[5]
-        + base4_encoded_slices[3] * BASE4_POWERS[6];
-
-    let low8 = low1 + bits1to8 * BASE4_POWERS[1];
-
-    let low61 = bits8to61 * BASE4_POWERS[8] + low8;
-
-    let low19 = low8
-        + base4_encoded_slices[4] * BASE4_POWERS[8]
-        + base4_encoded_slices[5] * BASE4_POWERS[16];
-
-    let high3 = base4_encoded_slices[11];
-
-    let high45 = bits19to61 + base4_encoded_slices[11] * BASE4_POWERS[42];
-
-    let high56 = bits8to61 + base4_encoded_slices[11] * BASE4_POWERS[53];
-
-    let high57 = high56 * BASE4_POWERS[1] + base4_encoded_slices[3];
-    let high58 = high57 * BASE4_POWERS[1] + base4_encoded_slices[2];
-
-    let high63 = base4_encoded_slices[1] + high58 * BASE4_POWERS[5];
-
-    let ror0 = low1 + high63 * BASE4_POWERS[1];
-    let ror1 = low1 * BASE4_POWERS[63] + high63;
-    let ror8 = low8 * BASE4_POWERS[56] + high56;
-    let ror19 = low19 * BASE4_POWERS[45] + high45;
-    let ror61 = low61 * BASE4_POWERS[3] + high3;
-    let rshift6 = high58;
-    let rshift7 = high57;
-    // 1, 8, 19, 61
-    // 6, 7
-
-    EncodedWitness { raw: w, w: ror0, ror1, ror8, ror19, ror61, rshift6, rshift7 }
+    // W2_POWER_SUM represents scalars that each base4_encoded_slices member must be multiplied by, in order to produce
+    // `encode(raw.rotate19) + encode(raw.rotate61) + encode(raw >> 6)`
+    let W2_POWER_SUM: [Field; 12] = [
```

Comment:
this is one of the key optimisations. When extending the message we need to compute the XOR of the witness rotated by 3 different values. The PR changes the algorithm so that this is one in a single linear expression, reducing the number of required additions